src/django/api/views/v1/opensearch_query_builder/production_locations_query_builder.py

```diff
@@ -140,4 +146,24 @@ def add_geo_bounding_box(self, top, left, bottom, right):
                     'right': right,
                 }
             }
-        }
+        })
+
+    def add_geo_polygon(self, values):
+        self.__init_filter()
+
+        geo_polygon_filter = None
+        for filter_item in self.query_body["query"]["bool"]["filter"]:
+            if "geo_polygon" in filter_item:
+                geo_polygon_filter = filter_item["geo_polygon"]
+                break
+
+        if geo_polygon_filter is None:
+            geo_polygon_filter = {"coordinates": {"points": []}}
+            self.query_body["query"]["bool"]["filter"].append({
+                "geo_polygon": geo_polygon_filter
+            })
+
+        points = geo_polygon_filter["coordinates"]["points"]
+        for point in values:
+            lat, lon = map(float, point.split(","))
+            points.append({"lat": lat, "lon": lon})
```

Comment:
Instead of mutating `geo_polygon_filter`, I would suggest to create a new version by parsing the `values` and constructing a new dictionary. We can then locate the filter in `self.query_body` and replace it with the new value. My point here is that mutating the object is much more riskier in terms of potential side effects and code clarity instead of replacing it.

Reply:
Good point, thank you. I've update `add_geo_polygon()` method to construct new `geo_polygon` dict and replace old value if necessary in `self.query_body`. In a nutshell:
1. Each request a new `geo_polygon` dict is created.
2. Remove previous `geo_polygon` from `self.query_body`.
3. Append newly created `geo_polygon` to `query_body["query"]["bool"]["filter"]`.

Note, that `self.query_body["query"]["bool"]` is still mulable dict for this and other builder methods. I apply immutability only to the `geo_polygon` here.

Reply:
^ It has been refactored a bit a but the main purpose remained the same - brand new `geo_polygon` wiil be created and attached to filter list.